core-rust/state-manager/src/accumulator_tree/storage.rs

```diff
@@ -0,0 +1,168 @@
+ * out of or in connection with (without limitation of any use, misuse, of any ledger system
+ * or use made or its functionality or any performance or operation of any code or protocol
+ * caused by bugs or programming or logic errors or otherwise);
+ *
+ * A. any offer, purchase, holding, use, sale, exchange or transmission of any
+ * cryptographic keys, tokens or assets created, exchanged, stored or arising from any
+ * interaction with the Work;
+ *
+ * B. any failure in a transmission or loss of any token or assets keys or other digital
+ * artefacts due to errors in transmission;
+ *
+ * C. bugs, hacks, logic errors or faults in the Work or any communication;
+ *
+ * D. system software or apparatus including but not limited to losses caused by errors
+ * in holding or transmitting tokens by any third-party;
+ *
+ * E. breaches or failure of security including hacker attacks, loss or disclosure of
+ * password, loss of private key, unauthorised use or misuse of such passwords or keys;
+ *
+ * F. any losses including loss of anticipated savings or other benefits resulting from
+ * use of the Work or any changes to the Work (however implemented).
+ *
+ * You are solely responsible for; testing, validating and evaluation of all operation
+ * logic, functionality, security and appropriateness of using the Work for any commercial
+ * or non-commercial purpose and for any reproduction or redistribution by You of the
+ * Work. You assume all risks associated with Your use of the Work and the exercise of
+ * permissions under this License.
+ */
+
+use std::collections::HashMap;
+
+/// The "read" part of an accumulator tree storage SPI.
+pub trait ReadableAccuTreeStore<N> {
+    /// Gets a vertical `TreeSlice` by its end index.
+    /// This index is effectively equal to the number of leaf nodes stored in the tree by _all_ the
+    /// slices _including_ the requested one.
+    fn get_tree_slice(&self, index: usize) -> Option<TreeSlice<N>>;
+}
+
+/// The "write" part of an accumulator tree storage SPI.
+pub trait WriteableAccuTreeStore<N> {
+    /// Puts a vertical `TreeSlice` at the given end index.
+    /// This index is effectively equal to the number of leaf nodes stored in the tree by _all_ the
+    /// slices _including_ the given one.
+    fn put_tree_slice(&mut self, index: usize, slice: TreeSlice<N>);
+}
+
+/// A convenience read+write storage trait.
+pub trait AccuTreeStore<N>: ReadableAccuTreeStore<N> + WriteableAccuTreeStore<N> {}
+impl<T: ReadableAccuTreeStore<N> + WriteableAccuTreeStore<N>, N> AccuTreeStore<N> for T {}
+
+/// A vertical slice of an accumulator tree.
+/// This is an "incremental" persistence part of a tree representing a batch of appended leaf nodes
+/// and the resulting merkle updates, propagating up to a single root.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct TreeSlice<N> {
+    /// The tree-levels of this slice, arranged from leafs to root.
+    pub levels: Vec<TreeSliceLevel<N>>,
```

Comment:
I'm not really convinced by this structure. It's easy to reason about when reading the definition of the struct, but when actually using it seems like a lot of headache. What do you think about a linearized representation of the whole tree?

Reply:
(just noting: we are still discussing options with Alexandru here)

In general, this follows the design from https://radixdlt.atlassian.net/wiki/spaces/RPNV1/pages/2996600937/RNP-7+-+Transaction+Hash+Tree+Inclusion+Exclusion+verification#Incremental-merkle-tree-storage which was created with following constraints in mind:

1. There is one tree per epoch header, but we need to persist its state to DB after each ledger header.
   - ideally we should achieve it by writing an amount of data proportional to the number of transactions in that commit (i.e. not proportional to number of transactions in that epoch so far)
2. The storage scheme should be tailored to level-like DBs (i.e. very-cheap inserts; less-cheap updates and deletes; no "append to existing entry" operation).
3. It would be nice to only read 1 (or some constant) number of DB entries during "tree update" and "generate proof" operations.